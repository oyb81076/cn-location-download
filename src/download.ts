import { loadCities, loadCounties, loadProvinces, loadTowns, loadVillages } from "./loader";
export interface IEntity {
  code: number;
  name: string;
  children?: IEntity[];
}
const loaders: Array<(url: string) => Promise<Array<{ code: number, name: string, link?: string }>>> = [
  loadProvinces, loadCities, loadCounties, loadTowns, loadVillages,
];

export async function download(url: string, index: number, depth: number, log = true): Promise<IEntity[]> {
  if (log) {
    console.log(`%s %s %s`, index, depth, url);
  }
  const data = await loaders[index](url);
  if (index < depth - 1) {
    const arr: IEntity[] = [];
    for (const { code, name, link } of data) {
      if (link) {
        arr.push({ code, name, children: await download(link, index + 1, depth) });
      } else {
        arr.push({ code, name });
      }
    }
    return arr;
  } else {
    return data.map(({ name, code }) => ({ name, code }));
  }
}
