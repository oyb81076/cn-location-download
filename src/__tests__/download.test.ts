import { download } from "../download";

test("depth=1", async () => {
  const data = await download("http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2017/index.html", 0, 1, false);
  expect(data).toMatchSnapshot();
});
