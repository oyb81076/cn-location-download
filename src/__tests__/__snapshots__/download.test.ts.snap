// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`depth=2 1`] = `
Array [
  Object {
    "code": 1100000000,
    "name": "北京市",
  },
  Object {
    "code": 1200000000,
    "name": "天津市",
  },
  Object {
    "code": 1300000000,
    "name": "河北省",
  },
  Object {
    "code": 1400000000,
    "name": "山西省",
  },
  Object {
    "code": 1500000000,
    "name": "内蒙古自治区",
  },
  Object {
    "code": 2100000000,
    "name": "辽宁省",
  },
  Object {
    "code": 2200000000,
    "name": "吉林省",
  },
  Object {
    "code": 2300000000,
    "name": "黑龙江省",
  },
  Object {
    "code": 3100000000,
    "name": "上海市",
  },
  Object {
    "code": 3200000000,
    "name": "江苏省",
  },
  Object {
    "code": 3300000000,
    "name": "浙江省",
  },
  Object {
    "code": 3400000000,
    "name": "安徽省",
  },
  Object {
    "code": 3500000000,
    "name": "福建省",
  },
  Object {
    "code": 3600000000,
    "name": "江西省",
  },
  Object {
    "code": 3700000000,
    "name": "山东省",
  },
  Object {
    "code": 4100000000,
    "name": "河南省",
  },
  Object {
    "code": 4200000000,
    "name": "湖北省",
  },
  Object {
    "code": 4300000000,
    "name": "湖南省",
  },
  Object {
    "code": 4400000000,
    "name": "广东省",
  },
  Object {
    "code": 4500000000,
    "name": "广西壮族自治区",
  },
  Object {
    "code": 4600000000,
    "name": "海南省",
  },
  Object {
    "code": 5000000000,
    "name": "重庆市",
  },
  Object {
    "code": 5100000000,
    "name": "四川省",
  },
  Object {
    "code": 5200000000,
    "name": "贵州省",
  },
  Object {
    "code": 5300000000,
    "name": "云南省",
  },
  Object {
    "code": 5400000000,
    "name": "西藏自治区",
  },
  Object {
    "code": 6100000000,
    "name": "陕西省",
  },
  Object {
    "code": 6200000000,
    "name": "甘肃省",
  },
  Object {
    "code": 6300000000,
    "name": "青海省",
  },
  Object {
    "code": 6400000000,
    "name": "宁夏回族自治区",
  },
  Object {
    "code": 6500000000,
    "name": "新疆维吾尔自治区",
  },
]
`;
